### DISCUSSION TITLE
commit verification using ssh key

### DISCUSSION THREAD

**Discussion by r2evans** [magit/magit#5046] [Ideas]
Code verification using the SSH key to sign commits is supported by both [GitHub](https://github.blog/changelog/2022-08-23-ssh-commit-verification-now-supported/) and [GitLab](https://docs.gitlab.com/ee/user/project/repository/signed_commits/ssh.html) (not sure if other git providers do). Is it possible now, and if not, what are the prospects of adding this functionality?

**Comment by tarsius** [comment]
Just follow the instructions in the second page you linked to and it will work.

**Comment by r2evans** [comment]
Okay, so `magit` supports it because `ssh` can do it automatically, is that the case? I think I was instead looking for an option to (en|dis)able it for each commit. (Sorry about the "obvious solution", I thought there would be more to it from the emacs/magit side.)

Thank you, and thanks for `magit`! It's a staple of my day-to-day work.

**Reply by tarsius** [reply]
> Okay, so `magit` supports it because `ssh` can do it automatically, is that the case?

Exactly!

> I think I was instead looking for an option to (en|dis)able it for each commit.

You should be able to set some variables to tell Git "*iff* we sign then use a ssh key" but still use the `--gpg-sign` to toggle whether you want to sign in this particular case.

The behavior of `magit-read-gpg-signing-key` used by `magit:--gpg-sign` (which implements handling the `--gpg-sign` argument in Magit), might be inconvenient when using ssh keys. This might need some work.

**Reply by algernon** [reply]
I was just researching how to do this, and made it work, by redefining `magit-read-gpg-signing-key`:

```emacs-lisp
(defun magit-read-gpg-signing-key (prompt &optional initial-input history)
  "")
```

It needs to return a truthy value, otherwise magit won't enable signing, and an empty string seems to do the trick, and will use the SSH signing key configured outside of magit. It's not a perfect solution, but it's a workaround that gets the job done.

**Reply by tarsius** [reply]
Thanks to caf37d0d608c14a8bca53ffa57675d5756e6c0b9, you don't have to redefine that function anymore.
